rllib/examples/centralized_critic_2.py

```diff
@@ -70,83 +70,72 @@ def value_function(self):
-        "opponent_obs": Discrete(6),
-        "opponent_action": Discrete(2),
-    })
+class FillInActions(DefaultCallbacks):
+    """Fills in the opponent actions info in the training batches."""
 
-    def __init__(self, env_config):
-        self.env = TwoStepGame(env_config)
-
-    def reset(self):
-        obs_dict = self.env.reset()
-        return self.to_global_obs(obs_dict)
-
-    def step(self, action_dict):
-        obs_dict, rewards, dones, infos = self.env.step(action_dict)
-        return self.to_global_obs(obs_dict), rewards, dones, infos
-
-    def to_global_obs(self, obs_dict):
-        return {
-            self.env.agent_1: {
-                "own_obs": obs_dict[self.env.agent_1],
-                "opponent_obs": obs_dict[self.env.agent_2],
-                "opponent_action": 0,  # populated by fill_in_actions
-            },
-            self.env.agent_2: {
-                "own_obs": obs_dict[self.env.agent_2],
-                "opponent_obs": obs_dict[self.env.agent_1],
-                "opponent_action": 0,  # populated by fill_in_actions
-            },
-        }
+    def on_postprocess_trajectory(self, worker, episode, agent_id, policy_id,
+                                  policies, postprocessed_batch,
+                                  original_batches, **kwargs):
+        to_update = postprocessed_batch[SampleBatch.CUR_OBS]
+        other_id = 1 if agent_id == 0 else 0
+        action_encoder = ModelCatalog.get_preprocessor_for_space(Discrete(2))
 
+        # set the opponent actions into the observation
+        _, opponent_batch = original_batches[other_id]
+        opponent_actions = np.array([
+            action_encoder.transform(a)
+            for a in opponent_batch[SampleBatch.ACTIONS]
+        ])
+        to_update[:, -2:] = opponent_actions
 
-def fill_in_actions(info):
-    """Callback that saves opponent actions into the agent obs.
 
-    If you don't care about opponent actions you can leave this out."""
+def central_critic_observer(agent_obs, **kw):
+    """Rewrites the agent obs to include opponent data for training."""
 
-    to_update = info["post_batch"][SampleBatch.CUR_OBS]
-    my_id = info["agent_id"]
-    other_id = 1 if my_id == 0 else 0
-    action_encoder = ModelCatalog.get_preprocessor_for_space(Discrete(2))
-
-    # set the opponent actions into the observation
-    _, opponent_batch = info["all_pre_batches"][other_id]
-    opponent_actions = np.array([
-        action_encoder.transform(a)
-        for a in opponent_batch[SampleBatch.ACTIONS]
-    ])
-    to_update[:, -2:] = opponent_actions
+    new_obs = {
+        0: {
+            "own_obs": agent_obs[0],
+            "opponent_obs": agent_obs[1],
+            "opponent_action": 0,  # filled in by FillInActions
+        },
+        1: {
+            "own_obs": agent_obs[1],
+            "opponent_obs": agent_obs[0],
+            "opponent_action": 0,  # filled in by FillInActions
+        },
+    }
+    return new_obs
 
```

Comment:
would this look different if you subclass the ObservationFunction?

Reply:
Probably just move it into __call__.